take a lock on a file before doing anything else so no 2 processes ca…

domsj opened the pull request:
…n be started for the same node

domsj committed:
take a lock on a file before doing anything else so no 2 processes can be started for the same node

toolslive commented:
- If someone kills the arakoon node with `SIGKILL` , the LOCK file will remain, 
  which will prevent future startups. Moreover, the stale LOCK file will contain a pid that hopefully does 
  not refer to a valid process (I can envision clever sysadmin writing scripts that do a something like
  `kill -9 $(cat LOCK)` ).
  I guess there's no foolproof solution to that. 
- Some deployers integrate with an init system (systemd, upstart, ....) that might have its own method 
  to ensure the service is only started once, so we should offer the option of not using this method.

domsj commented:
- A remaining lock will not prevent future startups. lockf/flock are not the infamous PID technique with which you continually seem to confuse it. It's a real lock that you can take on a file, enforced by the kernel.
  Below a small demonstration that it really works as it should:

```
... set up an arakoon ...
jan@jan-XPS-L412Z:~/dev/alba3$ cat /home/jan/dev/alba3/tmp/arakoon/nsm/nsm_0/LOCK 
Do not remove this file, it's meant to ensure an arakoon process is not running twice
jan@jan-XPS-L412Z:~/dev/alba3$ /home/jan/workspace/ARAKOON/arakoon/arakoon.native --node nsm_0 -config /home/jan/dev/alba3/tmp/arakoon/nsm.ini
problem: Could not get the required lock at /home/jan/dev/alba3/tmp/arakoon/nsm/nsm_0/LOCK, this node must already be running in another process
jan@jan-XPS-L412Z:~/dev/alba3$ pkill -9 arakoon
jan@jan-XPS-L412Z:~/dev/alba3$ ls /home/jan/dev/alba3/tmp/arakoon/nsm/nsm_0/
000.tlog  albamgr_plugin.cmxs  LOCK  nsm_0.db  nsm_0.db.wal  nsm_0.log  nsm_host_plugin.cmxs
jan@jan-XPS-L412Z:~/dev/alba3$ /home/jan/workspace/ARAKOON/arakoon/arakoon.native --node nsm_0 -config /home/jan/dev/alba3/tmp/arakoon/nsm.ini
jan@jan-XPS-L412Z:~/dev/alba3$ echo $?
42
```
- I've changed the code a bit so the lock file contains a friendly message instead of a PID, and added some more logic to prevent failing tests due to trying to take the same lock twice
- Even if they have an init script I still wouldn't make this configurable... It costs almost nothing to take the lock, and it still protects us against sysadmins trying to manually start an already running process. And it protects against bugs in their init system.

toolslive commented:
ok.
